nav2_map_server/src/map_3d/map_io_3d.cpp

```diff
@@ -0,0 +1,367 @@
+    if (pcd_origin.size() != 7) {
+      throw std::invalid_argument("view_point is wrong");
+    }
+
+    // Position
+    tf2::Vector3 translation = tf2::Vector3(
+      tf2Scalar(pcd_origin[0]),
+      tf2Scalar(pcd_origin[1]),
+      tf2Scalar(pcd_origin[2]));
+    // Orientation
+    tf2::Quaternion rotation = tf2::Quaternion(
+      tf2Scalar(pcd_origin[4]),
+      tf2Scalar(pcd_origin[5]),
+      tf2Scalar(pcd_origin[6]),
+      tf2Scalar(pcd_origin[3]));
+    // Transform
+    load_parameters.origin = tf2::Transform(rotation, translation);
+  } catch (YAML::Exception & e) {
+    std::cout << "[WARNING] [map_io_3d]: Couldn't load view_point from yaml file: " << std::endl;
+  }
+
+  return load_parameters;
+}
+
+void loadMapFromFile(
+  const LoadParameters & load_parameters,
+  sensor_msgs::msg::PointCloud2 & map_msg)
+{
+  sensor_msgs::msg::PointCloud2 map, map2;
+
+  std::cout << "[INFO] [map_io_3d]: Loading pcd_file: " <<
+    load_parameters.pcd_file_name << std::endl;
+
+  pcl::PCLPointCloud2::Ptr cloud(new pcl::PCLPointCloud2());
+
+  //  create a pcd reader for PointCloud2 data
+  pcl::PCDReader reader;
+
+  Eigen::Vector4f position;
+  Eigen::Quaternionf orientation;
+
+  // PCD version fixed to version 0.7
+  // will use sensor origin (added in PCD version 0.7)
+  int pcd_version = pcl::PCDReader::PCD_V7;
+
+  std::cout << "[DEBUG] [map_io_3d]: feeding pcd filename to reader" << std::endl;
+
+  if (reader.read(
+      load_parameters.pcd_file_name, *cloud,
+      position, orientation, pcd_version) == -1)              //* load the file
+  {
+    std::string error_msg{"Couldn't read "};
+    error_msg += load_parameters.pcd_file_name + "\n";
+    PCL_ERROR(error_msg.c_str());
+  }
+  std::cout << "[DEBUG] [map_io_3d]: pcd file is loaded" << std::endl;
+
+  std::cout << "[DEBUG] [map_io_3d]: converting pcd to message" << std::endl;
+
+  //  update message data
+  pclToMsg(map, cloud);
+
+  std::cout << "[DEBUG] [map_io_3d]: message conversion is done" << std::endl;
+
+  pcl_ros::transformPointCloud("random", load_parameters.origin, map, map2);
+  map_msg = map2;
+}
+
+LOAD_MAP_STATUS loadMapFromYaml(
+  const std::string & yaml_file,
+  sensor_msgs::msg::PointCloud2 & map_msg)
+{
+  if (yaml_file.empty()) {
+    std::cerr << "[ERROR] [map_io_3d]: YAML file name is empty, can't load!" << std::endl;
+    return LOAD_MAP_STATUS::MAP_DOES_NOT_EXIST;
+  }
+
+  std::cout << "[INFO] [map_io_3d]: Loading yaml file: " << yaml_file << std::endl;
+  LoadParameters load_parameters;
+
+  try {
+    std::cout << "initiating load_parameters";
+
+    load_parameters = loadMapYaml(yaml_file);
+  } catch (YAML::Exception & e) {
+    std::cerr << "[ERROR] [map_io_3d]: Failed processing YAML file " << yaml_file <<
+      " at position (" << e.mark.line << ":" << e.mark.column << ") for reason: " <<
+      e.what() << std::endl;
+
+    return LOAD_MAP_STATUS::INVALID_MAP_METADATA;
+  }
```

Comment:
`loadMapYaml()` can also generate an `std::exception` for some reasons (e.g. file operations will fail). So, there is also handling of this type of exceptions needed.